### ISSUE TITLE
Builder as installer for NPM scripts

### ISSUE THREAD

**Issue by ianobermiller** [FormidableLabs/builder#13]
I love the idea of archetypes and staying true to NPM scripts, but I'm not sold on using another tool / task runner. I'd be really interested in using a tool like builder that instead of replacing `npm run`, simply uses archetypes to add entries to my scripts section. The postinstall hooks that have been copied around, many inspired by Radium, are a great example of something that'd be nicer as an installable package. The nice thing about plain NPM scripts is that they are very approachable to newcomers, and don't require learning or installing anything new.

Nice work on pushing towards simpler builds/tasks, I'm psyched to see where this can go!

**Comment by ryan-roemer** [comment]
One of main thrusts here is really is "it's almost npm" for the scripts in the archetypes, etc.

The few things why there is I think a need for something beyond just `package.json` and files in the repo is a use case we're hitting in our OSS and client work:

* You have multiple (10-40) component repos that all have basically the same `package.json` scripts + config files.

Managing that via each repo is a pain, particularly when you find that one of your `package.json:scripts` commands has a bug and/or your config files need changes, etc.

Builder allows to you just take a dependency, manage expectations via semver and have "one source of truth" for common tasks across very common, but independent repositories.

Once you've got that use case, the things Builder offers are:

* Archetypes: The solution to the above problem. A way to define a specific set of tasks / configs / etc for one "type" of project. For us, that is the ever-expanding set of our https://github.com/FormidableLabs/?utf8=%E2%9C%93&query=victory repos (still have to roll this out, but it's our primary use case driver).
* Flexibility: Most of the meta tools I've used in the path are of the "buy the farm" nature. Work great when everything is within the workflow but fall apart once you want to be "just slightly" different. Builder solves this by allowing fine grain task overriding by name, where the larger composed tasks still stay the same and allow a specific repo's deviation from "completely off the shelf" to be painless.

It's also got the "nice to have" of a cross-OS compatible concurrent task runner (but you can get this as an independent node modules like `concurrently`, so not the biggest "thing" for a runner).

I think you're right in that Builder doesn't perhaps shine and isn't perhaps the best solution for:

* One-off repos. The project is specific enough and is the only thing that a group of folks manage.
* Beginning / learning node projects. One of the things we've just admitted for our Victory components is that a modern, publishable React build for these components + supporting lots of workflows + having a test driver infrastructure is **complicated**. And our build for these is meant to be efficient, robust and unfortunately, is not necessarily accessible in the straight build itself to completely new folks. (It _is_ expected that new folks / old folks / etc can easily _use_ Victory components in their projects with very basic workflows and builds...)

**Comment by ryan-roemer** [comment]
@ianobermiller  ... thinking about this for a little bit longer, I guess the alternatives / refactors to `builder` that would potentially suite our workflows:

* `npm` implements a plugin / module system to handle archetype-like scripts + dependencies. (But we're waiting on something in core `npm`...)
* We refactor builder to instead of using installed archetypes in `node_modules/ARCHETYPE/package.json` to instead _rewrite_ and mutate a projects `package.json` and other various files. (I think capturing what you might favor?) However, this becomes really tough and complicated once a project goes "off script" and wants to override just one `scripts` command when you consider the **upgrade** scenario for the archetype -- you're dealing with real code diffs + merges + potentially having an archetype upgrade blow away your customizations. Perhaps there is a sane way to manage the conflicts, but one first considering this thought this morning, seems... complicated and risky.

Love to hear your thoughts more, but perhaps you could aim to direct this issue into something actionable for builder (like a strawman refactor / rearchitecture that would be better) or close it?

Thanks!

**Comment by ianobermiller** [comment]
Apologies for the hit and run, I haven't taken the time to properly respond. I think I was not quite understanding the target use case of having an npm-installable set of common scripts for multiple projects that basically work the same (e.g. a bunch of plugins). I think that is the great sell for builder, and it is definitely missing in the first three introduction sentences, and only alluded to in the overview.

**Comment by colinmegill** [comment]
+1 for capturing this in the readme

**Comment by ryan-roemer** [comment]
> I think that is the great sell for builder, and it is definitely missing in the first three introduction sentences, and only alluded to in the overview.

Great point. Thanks @ianobermiller !

**Comment by ianobermiller** [comment]
I just saw [`butter-run-npm`](https://www.npmjs.com/package/better-npm-run), which gave me a simple idea that you could keep the npm scrips, and just point them to builder:

```js
"scripts": {
  "dist": "node ./node_modules/builder run dist"
}
```

Builder could even add these shims for you automatically. This would solve the main problem of yet-another-tool for end users.

**Comment by ryan-roemer** [comment]
@ianobermiller -- Yeah, the notion of doing environment variables + working with npm `scripts` is something that needs to be outside NPM for cross-platform because of the differences in env variables.

We're still working out whether or not (and how) to mutate an underlying project, and we're actually retooling the relationship of archetype `scripts`, `dependencies` and `devDependencies` to be more straightforward for consumers (https://github.com/FormidableLabs/builder/pull/18, https://github.com/FormidableLabs/builder-react-component/pull/6)

Also to hopefully fight the "yet another tool" issue, I'm currently writing up a section on "how to abandon Builder and go back to straight `npm`" as a guiding principle to keep it simple and aligned with `npm`
